sklearn/metrics/classification.py

```diff
@@ -292,6 +294,154 @@ def confusion_matrix(y_true, y_pred, labels=None, sample_weight=None):
+                             "multilabel classification.")
+
+        present_labels = unique_labels(y_true, y_pred)
+        if labels is None:
+            labels = present_labels
+            n_labels = None
+        else:
+            n_labels = len(labels)
+            labels = np.hstack([labels, np.setdiff1d(present_labels, labels,
+                                                     assume_unique=True)])
+
+        le = LabelEncoder()
+        le.fit(labels)
+        y_true = le.transform(y_true)
+        y_pred = le.transform(y_pred)
+        sorted_labels = le.classes_
+
+        # labels are now from 0 to len(labels) - 1 -> use bincount
+        tp = y_true == y_pred
+        tp_bins = y_true[tp]
+        if sample_weight is not None:
+            tp_bins_weights = np.asarray(sample_weight)[tp]
+        else:
+            tp_bins_weights = None
+
+        if len(tp_bins):
+            tp_sum = np.bincount(tp_bins, weights=tp_bins_weights,
+                                 minlength=len(labels))
+        else:
+            # Pathological case
+            true_sum = pred_sum = tp_sum = np.zeros(len(labels))
+        if len(y_pred):
+            pred_sum = np.bincount(y_pred, weights=sample_weight,
+                                   minlength=len(labels))
+        if len(y_true):
+            true_sum = np.bincount(y_true, weights=sample_weight,
+                                   minlength=len(labels))
+
+        # Retain only selected labels
+        indices = np.searchsorted(sorted_labels, labels[:n_labels])
+        tp_sum = tp_sum[indices]
+        true_sum = true_sum[indices]
+        pred_sum = pred_sum[indices]
+
+        fp = pred_sum - tp_sum
+        fn = true_sum - tp_sum
+        tp = tp_sum
+
+        if sample_weight is not None:
+            tn = sample_weight.sum() - tp - fp - fn
+        else:
+            tn = y_true.shape[0] - tp - fp - fn
+
+    else:
+        # check labels
+        if labels is not None:
+            max_label = y_true.shape[1] - 1
+            if np.max(labels) > max_label:
+                raise ValueError('All labels must be in [0, n labels). '
+                                 'Got %d > %d' %
+                                 (np.max(labels), max_label))
+            if np.min(labels) < 0:
+                raise ValueError('All labels must be in [0, n labels). '
+                                 'Got %d < 0' % np.min(labels))
+
+            n_labels = len(labels)
+            y_true = y_true[:, labels[:n_labels]]
+            y_pred = y_pred[:, labels[:n_labels]]
+
+        # account for sample weight
+        def _normal_or_sparse_row_multiply(A, w):
+            if issparse(A):
+                return A.multiply(w)
+            elif issparse(w):
+                return w.multiply(A)
+            else:
+                return np.multiply(A, w)
+
+        true_and_pred = _normal_or_sparse_row_multiply(y_true, y_pred)
```

Comment:
It's interesting that you found this faster than using indexing (i.e. indexing y_true with y_pred as a mask), which I'm sure was reported as faster in an earlier issue. I'm happy to leave it with this version for now at least, and we can worry about efficiency in a separate enhancement.

Reply:
I'm not sure what you mean by indexing, could you give an example? Thanks!
I found this to be faster than using labelencoder and bincount (both are quite expensive), and slightly faster than the [original implementation](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/classification.py#L1079-L1086), which is basically the same idea (I implemented this version because I didn't know about the original implementation before). It's not a huge improvement and made the code messier, I will change it back.

Reply:
Oh, I've gotten confused. I thought I used a different approach in my PR.

Reply:
If you want to know what I was talking about, see #8126. But that enhancement is probably best left for another PR.

Yes, cleaning up / minimising spurious variation between the current implementation and this PR might improve review.